Set HSF_PLATFORM to reasonable default

drbenmorgan opened the pull request:
As reported in Issue #6, template CPack file relies on HSF program
hsf_get_platform.py being present in the path. If it is not, the
platform variable HSF_PLATFORM is unset.

If HSF_PLATFORM is empty after an attempt to run hsf_get_platform.py,
set it to a value derived from CMake's knowledge of the system
processor arch, name and CXX compiler vendor ID. Do not use system
or compiler versions as these do not match the HSF conventions (e.g.
system version on Linux/Darwin is the kernel version).

Note that this assumes a CXX compiler has been configured.

drbenmorgan committed:
Set HSF_PLATFORM to reasonable default

As reported in Issue #6, template CPack file relies on HSF program
hsf_get_platform.py being present in the path. If it is not, the
platform variable HSF_PLATFORM is unset.

If HSF_PLATFORM is empty after an attempt to run hsf_get_platform.py,
set it to a value derived from CMake's knowledge of the system
processor arch, name and CXX compiler vendor ID. Do not use system
or compiler versions as these do not match the HSF conventions (e.g.
system version on Linux/Darwin is the kernel version).

Note that this assumes a CXX compiler has been configured.

peremato commented:
I am using the following code to 'emulate' for hsf_get_platform.py does (more or less)
```
function(GET_CONFIGURATION_TAG tag)
  #---arch--------------
  if(APPLE)
    execute_process(COMMAND uname -m OUTPUT_VARIABLE arch OUTPUT_STRIP_TRAILING_WHITESPACE)
  elseif(UNIX)
    execute_process(COMMAND uname -p OUTPUT_VARIABLE arch OUTPUT_STRIP_TRAILING_WHITESPACE)
  elseif(DEFINED ENV{Platform})
    set(arch $ENV{Platform})
    string(TOLOWER ${arch} arch)
  else()
    set(arch $ENV{PROCESSOR_ARCHITECTURE})
  endif()
  #---os----------------
  if(APPLE)
    set(os mac)
    execute_process(COMMAND sw_vers "-productVersion" 
                    COMMAND cut -d . -f 1-2 
                    OUTPUT_VARIABLE osvers OUTPUT_STRIP_TRAILING_WHITESPACE)
    string(REPLACE "." "" osvers ${osvers})
  elseif(WIN32)
    set(os win)
	execute_process(COMMAND cmd /c ver OUTPUT_VARIABLE osvers)
    if(osvers MATCHES "Version 5")
       set(osvers xp)  
    elseif(osvers MATCHES "Version 6")
       set(osvers 7)
    else()
       set(osvers)
    endif()
  else()
    execute_process(COMMAND cat /etc/issue OUTPUT_VARIABLE issue OUTPUT_STRIP_TRAILING_WHITESPACE)
    if(issue MATCHES Ubuntu)
      set(os ubuntu)
      string(REGEX REPLACE ".*Ubuntu ([0-9]+)[.].*" "\\1" osvers "${issue}")
    elseif(issue MATCHES SLC)
      set(os slc)
      string(REGEX REPLACE ".*release ([0-9]+)[.].*" "\\1" osvers "${issue}")
    else() 
      execute_process(COMMAND cat /etc/os-release OUTPUT_VARIABLE os_release OUTPUT_STRIP_TRAILING_WHITESPACE)
      if(os_release MATCHES "centos")
        set(os "centos")
        string(REGEX REPLACE ".*VERSION_ID=\"([0-9]+).*" "\\1" osvers "${os_release}")
      else()
        set(os linux)
        set(osvers)
      endif()
    endif() 
  endif()  
  #---compiler-----------
  if(DEFINED ENV{CC})
    set(compiler_cmd $ENV{CC})
  elseif(WIN32)
    find_program(compiler_cmd cl)
  else()
    find_program(compiler_cmd gcc)
  endif()
  if(compiler_cmd MATCHES cl.exe$)
    set(compiler vc)
    execute_process(COMMAND ${compiler_cmd} ERROR_VARIABLE versioninfo OUTPUT_VARIABLE out)
    string(REGEX REPLACE ".*Version ([0-9]+)[.].*" "\\1" cvers "${versioninfo}")
	math(EXPR cvers "${cvers} - 6")
  elseif(compiler_cmd MATCHES gcc[0-9]*$)
    set(compiler gcc)
    execute_process(COMMAND ${compiler_cmd} -dumpversion OUTPUT_VARIABLE GCC_VERSION)
    string(REGEX MATCHALL "[0-9]+" GCC_VERSION_COMPONENTS ${GCC_VERSION})
    list(GET GCC_VERSION_COMPONENTS 0 GCC_MAJOR)
    list(GET GCC_VERSION_COMPONENTS 1 GCC_MINOR)
    set(cvers ${GCC_MAJOR}${GCC_MINOR})
  elseif(compiler_cmd MATCHES icc$)
    set(compiler icc)
    execute_process(COMMAND ${compiler_cmd} -dumpversion OUTPUT_VARIABLE ICC_VERSION)
    string(REGEX MATCHALL "[0-9]+" ICC_VERSION_COMPONENTS ${ICC_VERSION})
    list(GET ICC_VERSION_COMPONENTS 0 ICC_MAJOR)
    set(cvers ${ICC_MAJOR})
  elseif(compiler_cmd MATCHES clang)
    set(compiler clang)
    execute_process(COMMAND ${compiler_cmd} -v ERROR_VARIABLE CLANG_VERSION)
    string(REGEX MATCHALL "[0-9]+" CLANG_VERSION_COMPONENTS ${CLANG_VERSION})
    list(GET CLANG_VERSION_COMPONENTS 0 CLANG_MAJOR)
    list(GET CLANG_VERSION_COMPONENTS 1 CLANG_MINOR)
    set(cvers ${CLANG_MAJOR}${CLANG_MINOR})
  else()
    set(compiler unk)
    set(cvers)
  endif()
  set(${tag} ${arch}-${os}${osvers}-${compiler}${cvers} PARENT_SCOPE)
endfunction()
```

drbenmorgan commented:
@peremato thanks! The OS part looks useful, so I'll look at adding that. For the arch and compiler parts, as the CPack file is in-project, it should know these already via `CMAKE_SYSTEM_PROCESSOR`, `CMAKE_C_COMPILER_ID` and `CMAKE_C_COMPILER_VERSION`. Is there a reason to re-query the system in the above code?

drbenmorgan committed:
Improve CMake implementation of platform tag

Provide function implementing major portion of `hsf_get_platform.py`
using pure CMake where possible:

- Derive compiler ID/version from CMake variables
  - Will use 'gnu' instead of 'gcc', though this follows general
    behaviour for naming via vendor rather than command
  - Will use 'appleclang' instead of `clang` as these may be
    distinct compilers on OS X (and it is impossible to derive
    a completely concrete Clang version from Apple's Clang).
- Platform info is *only relevant for host system*
  - Like hsf_get_platform, does not handle cross-compiling case yet
- On Linucies, prefer modern systems, falling back to older tools
  only when neccessary:
  1. /etc/os-release file
  2. lsb_release program
  3. Otherwise use what CMake knows.

Only use function if `hsf_get_platform.py` not available.

drbenmorgan commented:
The above commit provides a reasonably complete CMake implementation using ideas from Pere's code. 

One "feature" that should be noted is that all of these ways of deriving the platform tag will use information from the host, not target, system of the build so aren't safe for cross-compiling/packaging.
However, adding support for this is probably a simple case of testing the `CMAKE_CROSSCOMPILING` variable and using info from the toolchain file if that's the case.
